### DISCUSSION TITLE
[Chapter 09] Http 통신의 Keep-alive

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/spring-boot-for-msa#142] [Chapter 09. Application Integration: REST API] [answered]
P. 617 

> Http 1.0과 1.1 버전의 가장 큰 차이점은 커넥션 유지 기능이다. 
> Http 1.1 버전에서는 이를 Http persistent connection 혹은 keep-alive 라고 한다.

Q1. 책에서 keep-alive 에 대한 설명이 나오는데, Http 통신을 할 때 keep-alive 가 필요한 상황이 언제일까? 그리고 단점은 없을까?

Q2. Http 버전이 3.0까지 있는데 각 버전별 특징과 실제로 많이 쓰는 버전은?

**Comment by JoisFe** [comment, marked as answer]
<img width="756" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/90208100/0e6d5e05-6aba-4d92-b1b6-106ca01ab163">


- [참고자료](https://nginxstore.com/blog/nginx/keepalive-%EC%84%A4%EC%A0%95%EC%9C%BC%EB%A1%9C-nginx-http-%EC%84%B1%EB%8A%A5-%ED%96%A5%EC%83%81/)

**Comment by ahngilwoong** [comment]
# 검색 결과 !!!!


#### 1. HTTP 1.0 vs 1.1: Keep-Alive 기능

##### **HTTP 1.0**
- **특징**: '비연결성' 프로토콜.
- **작동 방식**: 각 요청과 응답 후 연결을 즉시 끊음.
- **결과**: 각 웹 페이지 요소(이미지, CSS 등)마다 새로운 TCP 연결 필요.

##### **HTTP 1.1 (Keep-Alive)**
- **특징**: '연결 유지' 기능.
- **작동 방식**: 연속된 요청과 응답을 동일한 연결에서 처리.
- **이점**: 연결 설정과 해제에 드는 시간과 자원 감소.

#### 2. Keep-Alive의 필요성 및 단점

##### **필요성**
- **적용 상황**: 
  - 웹 페이지가 다양한 리소스 로드 필요 시.
  - 이미지, CSS, JavaScript 파일 등에 대한 개별 연결 대신 기존 연결 재사용.
- **이점**: 
  - 페이지 로딩 속도 개선.
  - 서버와 클라이언트 간 통신 효율성 증가.

##### **단점**
- **서버 부담 증가**:
  - 동시에 여러 연결 유지로 인한 서버 연결 수 증가.
  - 서버 자원 부담 증가 가능성.
- **네트워크 자원 점유**:
  - 유휴 연결이 네트워크 자원 차지.
  - 다수 사용자 접속 환경에서 문제 가능성.

#### 3. 종합적인 평가
- **HTTP 1.1의 Keep-Alive 기능**은 웹 페이지 성능 개선에 크게 기여.
- **적절한 환경에서의 사용 중요**:
  - Keep-Alive 기능 적용 시 웹 페이지 복잡성, 서버 용량, 네트워크 환경 등 고려 필요.


참조 : 
[Keep-Alive](https://dabeen.medium.com/becoming-se-what-is-keep-alive-478b1ea99593)
[2](https://kamang-it.tistory.com/599)

**Comment by Irisation23** [comment]
### 금주 Check Point

- 해당 자료에 대해서는 같이 찾아보는 시간을 가졌음

**Comment by Irisation23** [comment]
https://mohwaproject.tistory.com/entry/Wire-Shake%EB%A5%BC-%ED%86%B5%ED%95%B4-%EB%B3%B8-KeepAlive-%EC%B2%98%EB%A6%AC-%EA%B3%BC%EC%A0%95
